pkg/cmd/pulumi/state_edit_frontend.go

```diff
@@ -0,0 +1,238 @@
+}
+
+type yamlStateFrontend struct {
+	header string
+}
+
+var _ snapshotFrontend = &yamlStateFrontend{}
+
+func (se *yamlStateFrontend) SnapshotToText(snap deploy.Snapshot) (snapshotText, error) {
+	dep, err := stack.SerializeDeployment(&snap, snap.SecretsManager, false)
+	if err != nil {
+		return "", err
+	}
+
+	b, err := json.Marshal(dep)
+	if err != nil {
+		return "", err
+	}
+
+	unk := map[string]interface{}{}
+	err = json.Unmarshal(b, &unk)
+	if err != nil {
+		return "", err
+	}
+
+	text := &bytes.Buffer{}
+
+	enc := yaml.NewEncoder(text)
+	enc.SetIndent(2)
+	if err = enc.Encode(unk); err != nil {
+		return "", fmt.Errorf("could not serialize deployment as YAML : %w", err)
+	}
+
+	sorted, err := sortYAML(text.Bytes())
+	if err != nil {
+		return "", err
+	}
+
+	return snapshotText(se.header + "\n" + string(sorted)), nil
+}
+
+func (se *yamlStateFrontend) TextToSnapshot(s snapshotText) (deploy.Snapshot, error) {
+	idk := map[string]interface{}{}
+	err := yaml.Unmarshal([]byte(s), &idk)
+	if err != nil {
+		return deploy.Snapshot{}, err
+	}
+
+	bJSON, err := json.Marshal(idk)
+	if err != nil {
+		return deploy.Snapshot{}, err
+	}
+
+	ctx := context.Background()
+	dep, err := stack.DeserializeUntypedDeployment(ctx, &apitype.UntypedDeployment{
+		Version:    3,
+		Deployment: bJSON,
+	}, stack.DefaultSecretsProvider)
+	if err != nil {
+		return deploy.Snapshot{}, err
+	}
+
+	if dep != nil {
+		return *dep, nil
+	}
+	return deploy.Snapshot{}, errors.New("could not deserialize deployment")
+}
+
+func sortYAML(doc []byte) ([]byte, error) {
+	var node yaml.Node
+	err := yaml.Unmarshal(doc, &node)
+	if err != nil {
+		return []byte{}, err
+	}
+	res := sortYAMLInner(&node)
+	b, err := yaml.Marshal(res)
+	return b, err
+}
+
+func sortYAMLInner(node *yaml.Node) *yaml.Node {
+	if node.Kind == yaml.DocumentNode {
+		for i, n := range node.Content {
+			node.Content[i] = sortYAMLInner(n)
+		}
+	}
+	if node.Kind == yaml.SequenceNode {
+		for i, n := range node.Content {
+			node.Content[i] = sortYAMLInner(n)
+		}
+	}
+	if node.Kind == yaml.MappingNode {
+		for i, n := range node.Content {
+			node.Content[i] = sortYAMLInner(n)
+		}
+		sort.Sort(nodes(node.Content))
+	}
+	return node
+}
+
+type nodes []*yaml.Node
+
+func (i nodes) Less(x, y int) bool {
+	x *= 2
+	y *= 2
+	xKey := i[x].Value
+	yKey := i[y].Value
+	priority := func(key string) int {
+		p, ok := map[string]int{
+			// Resource fields
+			"urn":                     0,
+			"custom":                  1,
+			"delete":                  2,
+			"id":                      3,
+			"type":                    4,
+			"inputs":                  5,
+			"outputs":                 6,
+			"parent":                  7,
+			"protect":                 8,
+			"external":                9,
+			"dependencies":            10,
+			"initErrors":              11,
+			"provider":                12,
+			"propertyDependencies":    13,
+			"pendingReplacement":      14,
+			"additionalSecretOutputs": 15,
```

Comment:
Two things:
First, this map definitely should not be inline here. It should be declared as a top-level lookup table.

More importantly I really don't like the idea of hand-writing these. If I add a thing in the middle, I'll have to renumber all these values. Consider changing this. There are two options:

- Use an `iota`-based enum. Those are auto-numbered, so you get to just specify the field order. You would then use [enumer](https://github.com/dmarkham/enumer) or [stringer](https://pkg.go.dev/golang.org/x/tools/cmd/stringer) to generate the string names. enumer will automatically generate a `<name>String(string) <name>` function. With stringer, you'll have to generate your own lookup table inside an `init()`. Either way, it'll be automatic. This will also obviate the need to do the 100 spacer since new fields can be added whenever and the rest will renumber. https://go.dev/play/p/J10gWozmRdy
- Do your own custom codegen.

I would suggest using enumer.